pkg/clusterapi/aws/actuator.go

```diff
@@ -0,0 +1,403 @@
+	// Extract cluster operator cluster
+	coCluster, err := a.clusterOperatorCluster(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	if coCluster.Spec.Hardware.AWS == nil {
+		return nil, fmt.Errorf("Cluster does not contain an AWS hardware spec")
+	}
+
+	region := coCluster.Spec.Hardware.AWS.Region
+	svc, err := a.createSVC(region)
+	if err != nil {
+		return nil, err
+	}
+
+	// For now, we store the machineSet resource in the machine.
+	// It really should be the machine resource, but it's not yet
+	// fleshed out in the Cluster Operator API
+	coMachineSet, coClusterVersion, err := a.clusterOperatorMachineSet(machine)
+	if err != nil {
+		return nil, err
+	}
+
+	if coClusterVersion.Spec.VMImages.AWSImages == nil {
+		return nil, fmt.Errorf("Cluster version does not contain AWS images")
+	}
+
+	// Get AMI to use
+	amiName := amiForRegion(coClusterVersion.Spec.VMImages.AWSImages.RegionAMIs, region)
+	if len(amiName) == 0 {
+		return nil, fmt.Errorf("Cannot determine AMI name from cluster version %#v and region %s", coClusterVersion, region)
+	}
+	imageIds := []*string{aws.String(amiName)}
+	describeImagesRequest := ec2.DescribeImagesInput{
+		ImageIds: imageIds,
+	}
+	describeAMIResult, err := svc.DescribeImages(&describeImagesRequest)
+	if err != nil {
+		return nil, fmt.Errorf("Error describing AMI %s: %v", amiName, err)
+	}
+	if len(describeAMIResult.Images) != 1 {
+		return nil, fmt.Errorf("Unexpected number of images returned: %d", len(describeAMIResult.Images))
+	}
+
+	// Describe VPC
+	vpcName := coCluster.Name
+	vpcNameFilter := "tag:Name"
+	describeVpcsRequest := ec2.DescribeVpcsInput{
+		Filters: []*ec2.Filter{&ec2.Filter{Name: &vpcNameFilter, Values: []*string{&vpcName}}},
+	}
+	describeVpcsResult, err := svc.DescribeVpcs(&describeVpcsRequest)
+	if err != nil {
+		return nil, fmt.Errorf("Error describing VPC %s: %v", vpcName, err)
+	}
+	if len(describeVpcsResult.Vpcs) != 1 {
+		return nil, fmt.Errorf("Unexpected number of VPCs: %d", len(describeVpcsResult.Vpcs))
+	}
+	vpcId := *(describeVpcsResult.Vpcs[0].VpcId)
+
+	// Describe Subnet
+	describeSubnetsRequest := ec2.DescribeSubnetsInput{
+		Filters: []*ec2.Filter{
+			&ec2.Filter{Name: aws.String("vpc-id"), Values: []*string{aws.String(vpcId)}},
+			&ec2.Filter{Name: aws.String("availability-zone"), Values: []*string{aws.String(defaultAvailabilityZone)}},
+		},
+	}
+	describeSubnetsResult, err := svc.DescribeSubnets(&describeSubnetsRequest)
+	if err != nil {
+		return nil, fmt.Errorf("Error describing Subnets for VPC %s: %v", vpcName, err)
+	}
+	if len(describeSubnetsResult.Subnets) != 1 {
+		return nil, fmt.Errorf("Unexpected number of Subnets: %d", len(describeSubnetsResult.Subnets))
+	}
+
+	// Determine security groups
+	var groupName, groupNameK8s string
+	if coMachineSet.Spec.Infra {
+		groupName = vpcName + "_infra"
+		groupNameK8s = vpcName + "_infra_k8s"
+	} else {
+		groupName = vpcName + "_compute"
+		groupNameK8s = vpcName + "_compute_k8s"
+	}
```

Comment:
We should give some thought to this for the future, been wondering if these implied names should be explicit in the API or not.